### DISCUSSION TITLE
Is it available for Mac M1?

### DISCUSSION THREAD

**Discussion by TheoBardu** [phoebe-team/phoebe#205] [Build help] [answered]
Hi to everyone. Excuse me for my probably trivial question, but I would like to know if someone has compiled Phoebe on a Mac with the M1 processor. 
If so, could you please tell me the steps?
Thank you a lot!

**Comment by jcoulter12** [comment]
Hi @TheoBardu, 

Yes, I've built Phoebe previously on my M1 Mac, so it should be possible. 

If I remember correctly, at the time I had some issues using OMP, though the situation may have improved, as I think the problem was mostly related to a conflict with the Eigen library. 

Have you attempted the build using the usual steps, and found that it failed? If so, can you show me the error? There's a good chance I can guess how to resolve it. 

Just as an additional note, if you want to run phonon-only calculations, (the lattice thermal conductivity using phonon-phonon scattering), an M1 Macbook may be powerful enough for this, though for production calculations you might have to wait a while. In general transport calculations, especially electron-phonon calculations, can be compute + memory intensive, so it's best to try to use a compute cluster or at least a good workstation. 

Regardless, happy to help you get the code working,
Jenny

**Reply by TheoBardu** [reply]
Hi Jenny! Thank you so much for your answer!
I want to compile Phoebe on my Mac M1  just as a post-processing tool. I'm going to do the heavy calculation steps on a supercomputer and then run Phoebe on mine. 

When I compile, different error occurs, and I can't understand what happen, since I should have installed all the packages required. I'm sorry if there are obvious errors, but I'm not so good in this.

I followed the steps on Phoebe page (https://phoebe.readthedocs.io/en/develop/installation.html) paying attention to the advices for installation on macOS systems. I've done the following steps:
**export LIBRARY_PATH=$LIBRARY_PATH:/opt/homebrew/Cellar/gcc/12.2.0/lib/gcc/12/
cmake -DOMP_AVAIL=OFF ../**
and occurred different errors ( I link the log, the outputs of cmake and the terminal output below)
[CMakeError.log](https://github.com/mir-group/phoebe/files/13578858/CMakeError.log)
[CMakeOutput.log](https://github.com/mir-group/phoebe/files/13578859/CMakeOutput.log)
[terminal_1.txt](https://github.com/mir-group/phoebe/files/13578894/terminal_1.txt)

Assuming there was a problem with mpi (which is strange because I have mpif90 installed), I tried to do so:
**export LIBRARY_PATH=$LIBRARY_PATH:/opt/homebrew/Cellar/gcc/12.2.0/lib/gcc/12/
cmake -DOMP_AVAIL=OFF -DMPI_AVAIL=OFF -DHDF5_AVAIL=OFF ..**
 but again, these are the results:
[CMakeError.log](https://github.com/mir-group/phoebe/files/13578936/CMakeError.log)
[CMakeOutput.log](https://github.com/mir-group/phoebe/files/13578937/CMakeOutput.log)
[terminal_2.txt](https://github.com/mir-group/phoebe/files/13578942/terminal_2.txt)


If you can help me, I appreciate a lot! 
Thank you so much! Have a nice day!
Theo

**Reply by jcoulter12** [reply, marked as answer]
Hi Theo, 

That sounds like a good plan, let me see if I can help you get set up. 

Your first error about kokkos-kernels is easy to fix, you probably missed the "recurse-submodules" part of the git clone line that's on the installation page. You can fix that by just running: `git submodule update --init`

Now, for the main problem -- this was actually somewhat tricky, some recent changes to the Mac environment made this build harder than I suspected. Thanks for bringing it to my attention, I wish Macs were as code-friendly as they used to be. 

I eventually got the build bellow working on my laptop -- would you mind checking that this works for you also? 

```
# if you don't have these already, use homebrew to install: 

brew install cmake
brew install gcc
brew install llvm
# stop here and follow the instructions llvm provides regarding setting path variables -- run whatever line it tells you that looks like the one below, where username should be your laptop username. 
#echo 'export PATH="/opt/homebrew/opt/llvm/bin:$PATH"' >> /Users/username/.bash_profile

brew install open-mpi
brew install hdf5-mpi
brew install scalapack
brew install libomp

# before building, set these flags
export CC=/opt/homebrew/opt/llvm/bin/clang
export CXX=/opt/homebrew/opt/llvm/bin/clang++
export LDFLAGS="-L/opt/homebrew/opt/llvm/lib/c++ -Wl,-no_compact_unwind,-rpath,/opt/homebrew/opt/llvm/lib/c++"
export CXXFLAGS="-I/opt/homebrew/opt/llvm/include -I/opt/homebrew/opt/libomp/include -fopenmp"
export CFLAGS="-I/usr/local/opt/libomp/include -I/opt/homebrew/opt/llvm/include"
export LIBRARY_PATH=$LIBRARY_PATH:/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/
export SDKROOT=$(xcrun --show-sdk-path)

mkdir build
cd build
cmake ../
make -j 4 phoebe
```

Two extra notes -- 
1) You should make sure that "/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/" points to an actual path on your laptop, and that this is the correct path to the gcc homebrew installed. 
But, you'll need to replace "username" with 
2) There's some issue with llvm that means exceptions cannot be properly caught in c++ codes. There's only one way this could be an issue in Phoebe, and if you run into it, you'll see an error like: 
`libc++abi: terminating due to uncaught exception of type HighFive::DataSetException: Unable to open the dataset "/physical_unit": (Symbol table) Object not found`
It's not something I can fix, as it seems like it's due to a bug in a dependency. Therefore, if this happens, just let me know, there should be an easy work around. 

Let me know how it goes, and if this works I'll update the docs. 
Thanks again, 
Jenny

**Reply by TheoBardu** [reply]
Thank you a lot Jenny.
I have done all what you said , but unfortunately errors occurred.
I link you again the terminal page and the log files.

[terminal.txt](https://github.com/mir-group/phoebe/files/13623381/terminal.txt)
[CMakeError.log](https://github.com/mir-group/phoebe/files/13623382/CMakeError.log)
[CMakeOutput.log](https://github.com/mir-group/phoebe/files/13623383/CMakeOutput.log)


I have installed the Phoebe code from scratch from git, using the command "**git clone --recurse-submodules https://github.com/mir-group/phoebe.git**".
I checked for the "/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/" directory and it points to a path. 
I installed from brew all the libraries that you listed me, using the command **echo 'export PATH="/opt/homebrew/opt/Ilvm/bin:$PATH" >> ~/.zshrc'** as it suggested me after the installation of llvm.

I used all the export setting for the flags, but after all, it give me the error "Could NOT find MPI_C" and the others that you can see from the files I linked above. I checked all the export and all the paths exists. 

I'm sorry if there is something I'm doing wrong. 
Thank you so much for all your support.
Let me know if there is something I can do.
Thanks, 
Theo

**Reply by jcoulter12** [reply]
Hi Theo, 

No worries, thanks for working through this with me and being patient. 

`brew install open-mpi`

Can you check that this was successful, and that you have file existing in a location something like: 
-- Found MPI_C: /opt/homebrew/Cellar/open-mpi/5.0.0/lib/libmpi.dylib (found version "3.1") 
-- Found MPI_CXX: /opt/homebrew/Cellar/open-mpi/5.0.0/lib/libmpi.dylib (found version "3.1") 
-- Found MPI_Fortran: /opt/homebrew/Cellar/open-mpi/5.0.0/lib/libmpi_usempif08.dylib (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  

I didn't have to set any flags related to this. 

One other thing, are you removing the contents of your build directory before running cmake again? 
Occasionally I forget to do this and it causes me issues. 

Let me know, 
Jenny

**Reply by TheoBardu** [reply]
Hi Jenny, 
I just checked the path and all the libraries are present in the directory **/opt/homebrew/Cellar/open-mpi/5.0.0/lib**.

<img width="1284" alt="Screenshot 2023-12-10 alle 09 43 18" src="https://github.com/mir-group/phoebe/assets/112978812/e5bde38a-d67b-4b23-93cd-fd79d5adaf10">

Yes, anytime I want to do the procedure with cmake, I delete the build directory and I start again.

**Reply by jcoulter12** [reply]
Hi Theo, 

Ok, thanks for confirming all of this. A few more suggestions based on a closer look at the output files you've helpfully provided -- 

1) I took a diff of our terminal outputs and the only thing that's different is mine finds OMP 5.1 and yours finds OMP 4.5. I don't suspect that's an issue, but it makes me wonder if maybe you have some packages which are outdated? 
My laptop is quite new right now, so for me everything was a very fresh installation, so everything was a new install. You can check this by running: 
`brew update` to list out of date packages and then `brew upgrade` if you want to update anything. 

2) However, my key suspicion right now is that anaconda is on your path and may be obscuring things (a super common issue). 
I notice in your `CMakeError.log` file: 
` -I/Users/theo/opt/anaconda3/include` and lib are in your CMake env variables. For example, it's looking at files like: 
` /Users/theo/opt/anaconda3/lib/libpmpi.dylib `
 and throwing errors such as:
` Fatal Error: Cannot read module file '/Users/theo/opt/anaconda3/include/mpi.mod' opened at (1), because it was created by a different version of GNU Fortran`

There are a few things you can do to try to figure out where this is coming from:

  * Make sure you don't have a conda env activated. 
  
  * I'm making one edit to my above instructions, I'm changing: 
  `export CFLAGS="$CFLAGS -I/usr/local/opt/libomp/include -I/opt/homebrew/opt/llvm/include"`
  to 
  `export CFLAGS="-I/usr/local/opt/libomp/include -I/opt/homebrew/opt/llvm/include"`
  So that any previously established flags won't be on that path. It was a typo on my part. 
  
  * Check what your env thinks it has for MPI compilers, by typing `which MPICXX`, `which MPIF90`, and `which MPICC`, to confirm these are set to things like: 
  `/opt/homebrew/bin/MPICC`
  If not, force them to use the homebrew copies by checking that these MPI* files exist in the right place, then setting also the flags: 
  `export MPIF90=/opt/homebrew/bin/MPIF90`
  `export MPICC=/opt/homebrew/bin/MPICC`
  `export MPICXX=/opt/homebrew/bin/MPICXX`
  and see if that helps anything. 

Thanks, 
Jenny

**Reply by TheoBardu** [reply]
Thank you so much. 
I upgraded with **brew upgrade** formula all the library. 
I have disabled anaconda in order to avoid the conflict using only the homebrew installed libraries. 
The cmake command was successfully executed using all the export setups you indicated me. 
It seems now Phoebe works. I'm going to perform some calculations and some tests to verify if it works.

I really don't know how to thank you. 
I apologize for the waste of time.
Thank you again. 
Greetings,
Theo

**Reply by jcoulter12** [reply]
Hi Theo, 

No worries, thanks again for pointing out the difficulty of doing this build, I'm going to add these instructions to do the installation page now that we have confirmed they work. Thanks for helping to confirm these are indeed the right instructions. 

I'm happy to answer future questions about the code if you encounter them -- it's still a fairly new code, so we're glad to have feedback. 
On a separate note, I also have a couple of recent bugfixes I will push to the main branch of the code soon, I'll let you know once this is done, and ask that you pull from github to get them. 

Best, 
Jenny
